### PR TITLE
Refactor Error handling to allow multiple errors and report location

### LINKED ISSUE #18: Refactor `ParserError` to be a `struct`.
Currently our `ParserError` is an `enum` with a single `SyntaxError` variant:

https://github.com/partiql/partiql-lang-rust/blob/4aa104c65ee7b69eca967c10e823016e00c91493/partiql-parser/src/result.rs#L130-L136

More likely than not, we all `ParserErrors` will have a message with optional position information, we should refactor the variant into some kind of `ParserErrorType` or something that could be as simple as just the tag (e.g. `SyntaxError`), but could also have error specific data.

See also #16, no matter what we do here, we'll like a utility method to do non-message equality.

### PR DESCRIPTION
**Pull request by jpschorr**
Work towards #77 
Fixes #18 
Maybe all or part of #14 

Refactor Error handling to allow multiple errors and report location

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-parser/src/lalr/mod.rs [hunk 1]
```diff
@@ -29,31 +29,74 @@ mod grammar {
...
 pub use lexer::Spanned;
 pub use lexer::Token;
 
-type LexerError = (ByteOffset, lexer::LexError, ByteOffset);
-type LalrpopError<'input> = ParseError<ByteOffset, lexer::Token<'input>, LexerError>;
-pub type ParserResult<'input> = Result<Box<ast::Expr>, ParserError<'input>>;
+type LalrpopError<'input> =
+    ParseError<ByteOffset, lexer::Token<'input>, ParserError<'input, BytePosition>>;
+type LalrpopResult<'input> = Result<Box<ast::Expr>, LalrpopError<'input>>;
+type LalrpopErrorRecovery<'input> =
```
Comment (alancai98) [inline partiql-parser/src/lalr/mod.rs:42]: Oh interesting. Didn't know lalrpop had some notion of [error recovery](http://lalrpop.github.io/lalrpop/tutorial/008_error_recovery.html). I'm curious as to what other parse errors we could make recoverable.

File: partiql-parser/src/result.rs [hunk 2]
```diff
@@ -2,38 +2,125 @@
...
+    }
+}
+
+pub type LexicalError<'input, L> = Located<LexError<'input>, L>;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct UnexpectedTokenData<'input> {
+    /// The unexpected token
+    pub token: Token<'input>,
+    // TODO expected: ...,
```
Comment (alancai98) [inline partiql-parser/src/result.rs:63]: nit: could add an issue to track adding `expected`. And could reference at this TODO: https://github.com/partiql/partiql-lang-rust/blob/dcef86525302fdc990688b8c2220717d92d22376/partiql-parser/src/lalr/mod.rs#L91

**Review by alancai98 — APPROVED** [review]
Looks good. Just had a few questions and comments about creating tracking issues.

File: partiql-parser/src/location.rs [hunk 3]
```diff
@@ -175,28 +175,58 @@ impl fmt::Display for LineAndColumn {
...
     }
 }
 
-impl fmt::Display for Position {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self.0 {
-            None => write!(f, "unknown position"),
-            Some(location) => write!(f, "{}", location),
-        }
+impl<T, Loc: Debug> ToLocated<Loc> for T {}
```
[the remark this answers is not in the document]
Comment (jpschorr) [reply partiql-parser/src/location.rs:206]: It is needed; I'll add docs.

It is a "blanket implementation" (see https://doc.rust-lang.org/book/ch10-02-traits.html#using-trait-bounds-to-conditionally-implement-methods) for any type.

Essentially, it auto-implements the `ToLocated` trait for all types in a mod that `uses` `ToLocated`. It's why the code can call the `to_located` on strings & structs and such in the `tests` mod at the bottom of this file.

File: partiql-parser/src/lalr/mod.rs (hunk 1 above)
Comment (jpschorr) [reply partiql-parser/src/lalr/mod.rs:42]: Yeah. This PR add it to `TableReference` (essentially for each thing that can be a "FROM source").

I think we should likely explore other such failure points for each possible clause of SFW, and liley for ExprTerm as well.

It'll be nice to be able to catch multiple parse errors per parse.

Comment (jpschorr) [reply partiql-parser/src/lalr/mod.rs:42]: See #88

File: partiql-parser/src/result.rs (hunk 2 above)
Comment (jpschorr) [reply partiql-parser/src/result.rs:63]: Added #89

File: partiql-parser/src/result.rs [hunk 4]
```diff
@@ -2,38 +2,125 @@
...
 }
 
+impl<'input, Loc: Debug> ParserError<'input, Loc>
+where
+    Loc: Debug, // TODO this should be `Loc: Display`
+{
+    /// Maps an `ParserError<Loc>` to `ParserError<Loc2>` by applying a function to each variant
+    pub fn map_loc<F, Loc2>(self, tx: F) -> ParserError<'input, Loc2>
+    where
+        Loc2: Debug, // TODO this should be `Loc2: Display`
```
[the remark this answers is not in the document]
Comment (jpschorr) [reply partiql-parser/src/result.rs:45]: Added #90

File: partiql-parser/src/location.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-parser/src/location.rs:207]: Ah got it. Forgot about "blanket implementation"s.

**Pull request merged**